srcts/src/components/darkModeSwitch.ts

```diff
@@ -0,0 +1,312 @@
+
+    if (typeof this.mode === "undefined") {
+      this.mode = window.matchMedia("(prefers-color-scheme: dark)").matches
+        ? "dark"
+        : "light";
+    }
+
+    this.reflectPreference();
+
+    // Sync with system changes
+    window
+      .matchMedia("(prefers-color-scheme: dark)")
+      .addEventListener("change", ({ matches: isDark }) => {
+        this.mode = isDark ? "dark" : "light";
+        this.reflectPreference();
+      });
+
+    this.observeDocumentThemeAttribute();
+  }
+
+  observeDocumentThemeAttribute(): void {
+    const observer = new MutationObserver((mutations) => {
+      mutations.forEach((mutation) => {
+        if (mutation.target !== document.documentElement) return;
+        if (mutation.attributeName !== this.attribute) return;
+        const newValue = document.documentElement.getAttribute(this.attribute);
+        if (!newValue || newValue === this.mode) return;
+        this.mode = newValue as DarkModeSwitch["mode"];
+      });
+    });
+
+    const config = {
+      attributes: true,
+      childList: false,
+      subtree: false,
+    };
+
+    observer.observe(document.documentElement, config);
+  }
+
+  getValue(): string {
+    return this.mode;
+  }
+
+  render(): ReturnType<LitElement["render"]> {
+    const other = this.mode === "light" ? "dark" : "light";
+    const label = `Switch from ${this.mode} to ${other} mode`;
+
+    return html`
+      <button
+        title="${label}"
+        aria-label="${label}"
+        aria-live="polite"
+        data-theme="${this.mode}"
+        @click="${this.onClick}"
+      >
+        <svg class="sun-and-moon" aria-hidden="true" viewBox="0 0 24 24">
+          <mask class="moon" id="moon-mask">
+            <rect x="0" y="0" width="100%" height="100%" fill="white" />
+            <circle cx="25" cy="10" r="6" fill="black" />
+          </mask>
+          <circle
+            class="sun"
+            cx="12"
+            cy="12"
+            r="6"
+            mask="url(#moon-mask)"
+            fill="currentColor"
+          />
+          <g class="sun-beams" stroke="currentColor">
+            <line x1="12" y1="1" x2="12" y2="3" />
+            <line x1="12" y1="21" x2="12" y2="23" />
+            <line x1="4.22" y1="4.22" x2="5.64" y2="5.64" />
+            <line x1="18.36" y1="18.36" x2="19.78" y2="19.78" />
+            <line x1="1" y1="12" x2="3" y2="12" />
+            <line x1="21" y1="12" x2="23" y2="12" />
+            <line x1="4.22" y1="19.78" x2="5.64" y2="18.36" />
+            <line x1="18.36" y1="5.64" x2="19.78" y2="4.22" />
+          </g>
+        </svg>
+      </button>
+    `;
+  }
+
+  receiveMessage(el: DarkModeSwitch, data: DarkModeMessageToggle): void {
+    if (data.method === "toggle") {
+      if (typeof data.value === "undefined" || data.value === null) {
+        data.value = this.mode === "light" ? "dark" : "light";
+      }
+      el.setAttribute("mode", data.value);
+    }
+  }
```

Comment:
I think I'd prefer just to remove this (and consequently `DarkModeMessageToggle` as well).

I guess I can also see an argument for keeping it, which I'd be fine with, but in that case, we should add some commentary about why it's currently unused

```suggestion
```

Reply:
I came up with a slightly different approach: just like we store `isShinyInput` as a static property, I added `shinyCustomMessageHandlers` as a static property as well. Then in `webComponents`, after making the input binding from the class, we can also register any custom message handlers.

I like that this keeps the message handler implementation within the class definition, so outside code only needs to know how to add the message handlers. And that part logically happens next to the input binding registration...

See b4f0e9c222a1685ba0577694bc0294122e7d625b for the changes

Reply:
Makes sense, thanks.